rfcs/20190430-tokenization-conventions.md

```diff
@@ -0,0 +1,352 @@
+# RFC: Tokenization API Conventions & Initial Implementations
+
+Status        | Proposed
+:------------ | :-----------------------------------
+**Author(s)** | Robby Neale (Google)
+**Sponsor**   | Mark Omernick (Google), Greg Billock (Google)
+**Updated**   | 2019-04-18
+
+## Objective {#objective}
+
+Establish a set of principles to follow that will unify Tensorflow tokenization
+offerings, and apply these conventions to APIs for a set of op-level tokenizers.
+
+## Motivation {#motivation}
+
+There are a number of tokenization methods we wish to make available for
+converting text runs into sequenced substrings processed by Tensorflow graphs.
+In the past, these steps needed to be performed outside the graph in a data
+preprocessing step, or through custom ops. The former had a chance of creating
+skew if the preprocessing wasn't performed consistently, and the latter
+fragmented Tensorflow NLP usage.
+
+## User Benefit {#user-benefit}
+
+To prevent further fragmentation, and to the benefit of all NLP modelers, we
+wish to establish a set of API conventions for tokenization ops to follow that
+will make them easy to use, switch between, and compose. There is not one
+best tokenizer for all use cases, and it is not a goal to establish this best
+tokenizer or a single API for all tokenizers to rigidly follow. [[See appendix A
+for further details](#appendix-a-conventions-vs-oo)]
+
+In addition to conventions to follow for tokenization ops, we intend to
+discuss new APIs for whitespace split, Unicode script split, and wordpiece.
+
+## Design Proposal {#design-proposal}
+
+The base API for each tokenizer is that it takes a Tensor or
+[RaggedTensor](https://www.tensorflow.org/guide/ragged_tensors) of strings (or
+optionally integer Unicode code points) as input, and outputs a RaggedTensor of
+tokens. Tokenizers can come in the form of ops, subgraphs, TF.Hub modules, or TF
+Serving models. Tokens can be strings or integers (frequently as vocabulary
+indices). By standardizing on inputs and output formats, we aim to allow for
+ease in composability between tokenizers (see example of this in the
+custom_tokenizer example below).
+
+Below is a list of guidelines for the tokenization APIs, followed by
+explanations regarding the reasoning behind each choice, and finally we list out
+specific tokenization APIs that we wish to make available first.
+
+### Conventions {#conventions}
+
+1.  Tokenizers should be able to accept a Tensor or RaggedTensor of UTF-8
+    encoded strings as input.
+1.  Tokenizers should output a RaggedTensor of tokens that includes an added
+    ragged dimension of tokens for each input string, unless the input is a
+    scalar, in which case the output will be a vector Tensor.
+1.  Tokenizers should also provide a `_with_offsets` variant that includes start
+    and limit offsets with the output which will allow alignment of the tokens
+    to the original byte offsets of the strings they are tokenized from.
```

Comment:
From the examples I can see that the output token is not required to be byte equal to the original input substring. Perhaps it is worth making this explicit here (since not all users of tokenizers may know about NFC normalization etc).

Reply:
Agreed. I'll include this with the interface update.